core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -681,6 +694,17 @@ class GroupCoordinator(val brokerId: Int,
   def onCompleteJoin(group: GroupMetadata) {
     var delayedStore: Option[DelayedStore] = None
     group synchronized {
+      if (group.generationId == 0) {
+        if (group.newMemberAdded && time.milliseconds() < group.initialRebalanceTimeout) {
+          group.newMemberAdded = false
+          val delay = scala.math.min(groupConfig.groupInitialRebalanceDelayMs, group.initialRebalanceTimeout - time.milliseconds()).toInt
+          val groupKey = GroupKey(group.groupId)
+          joinPurgatory.watchOperation(new DelayedJoin(this, group, delay, group.initialRebalanceTimeout), Seq(groupKey))
```

Comment:
@dguy thinking a bit more on the two different delay extension policy, I now actually feel in practice maybe the slightly-modified one will work better. Because with the default values (10, and 3 seconds), it means that with 4 or more consumer members within a group, the original policy will always drag the prepare-phase to 10 second even if they all join at a small time range, say from time 0 to 3; whereas the new mechanism is worse if you have fewer than `rebalance / delay` members, which sounds less likely. So maybe we can just follow the newer mechanism and clarify that in the wiki.